### PR TITLE
Us030

### PR DESCRIPTION
**Pull request by rjweeks**
View all collection exercises for an RU and Survey.

If an Respondent Enrolment event, check the state of all collection exercises and transition if required.

### CODE REVIEW

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java [hunk 1]
```diff
@@ -185,21 +188,76 @@ public CaseEvent createCaseEvent(CaseEvent caseEvent, Case newCase, Timestamp ti
...
+    checkCaseState(category, caseGroups, caseEvent, newCase);
+  }
+
+  /**
+   * This has been triggered by a 'RESPONDENT_ENROLED' event. If we find any associated Case Groups
+   * with only B cases we need to make case 'INACTIONABLE' and create BI case.
+   */
+  private void checkCaseState(final Category category, List<CaseGroup> caseGroups, CaseEvent caseEvent, Case newCase) throws CTPException {
+    // check all case groups are type B. For surveys this is always true
+    List<CaseGroup> caseGroupsToUpdate = caseGroups.stream().filter(cg -> cg.getSampleUnitType().toString().equals("B")).collect(Collectors.toList());
```
Comment (andrewmil) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java]: the `toString()` in this line is not required

File: src/test/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImplTest.java [hunk 2]
```diff
@@ -1185,6 +1182,19 @@ public void testEventRespondentEnrolled() throws Exception {
...
+    // new mocks
+
+    List<CollectionExerciseDTO> listCollex = Collections.singletonList(makeCollectionExercise());
+    when(collectionExerciseSvcClientService.getCollectionExercises(null)).thenReturn(listCollex);
+    List<CaseGroup> caseGroups = Collections.singletonList(makeCaseGroup());
+    CaseGroup caseGroup = makeCaseGroup();
+    when(caseGroupRepo.retrieveByPartyIdInListOfCollEx(any(), any())).thenReturn(caseGroups);
+    when(caseGroupRepo.findOne(ENROLMENT_CASE_INDIVIDUAL_FK)).thenReturn(caseGroup);
+    List<Case> c = Collections.singletonList(makeCase());
+    when(caseRepo.findByCaseGroupFKOrderByCreatedDateTimeDesc(any())).thenReturn(c);
```
Comment (andrewmil) [inline src/test/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImplTest.java]: Do we also needs to add some new assertions for the new functionality?

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java [hunk 3]
```diff
@@ -185,21 +188,76 @@ public CaseEvent createCaseEvent(CaseEvent caseEvent, Case newCase, Timestamp ti
...
+    CollectionExerciseDTO collectionExercise = collectionExerciseSvcClientService
+            .getCollectionExercise(caseGroup.getCollectionExerciseId());
+    // fetch all the collection exercises for a survey
+    List<CollectionExerciseDTO> collectionExercises = collectionExerciseSvcClientService.getCollectionExercises(collectionExercise.getSurveyId());
+    // get published collection exercise
+    List<CollectionExerciseDTO> publishedCollexs = collectionExercises.stream().filter(ce -> ce.getState().toString().equals("READY_FOR_LIVE")).collect(Collectors.toList());
+    // get list of collection exercise ids
+    List<UUID> collExs = publishedCollexs.stream().map(CollectionExerciseDTO::getId).collect(Collectors.toList());
+    // fetch party ID for the RU
+    UUID partyId = targetCase.getPartyId();
```
[the remark this answers is not in the document]
Comment (andrewmil) [reply src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java]: Ignore that I forgot that for B-cases the partyId represents a business party id. That can get a bit confusing.

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java (hunk 1 above)
Comment (rjweeks) [reply src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java]: I think it is to get a proper String comparison.

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java [hunk 4]
```diff
@@ -185,21 +188,76 @@ public CaseEvent createCaseEvent(CaseEvent caseEvent, Case newCase, Timestamp ti
...
+   * with only B cases we need to make case 'INACTIONABLE' and create BI case.
+   */
+  private void checkCaseState(final Category category, List<CaseGroup> caseGroups, CaseEvent caseEvent, Case newCase) throws CTPException {
+    // check all case groups are type B. For surveys this is always true
+    List<CaseGroup> caseGroupsToUpdate = caseGroups.stream().filter(cg -> cg.getSampleUnitType().toString().equals("B")).collect(Collectors.toList());
+    for(CaseGroup cg : caseGroupsToUpdate) {
+      // fetch cases associated to case group
+      List<Case> cases = caseRepo.findByCaseGroupFKOrderByCreatedDateTimeDesc(cg.getCaseGroupPK());
+      List<Case> bCases = cases.stream().filter(c -> c.getSampleUnitType().toString().equals("B")).collect(Collectors.toList());
+      //List<Case> biCases = cases.stream().filter(c -> c.getSampleUnitType().toString().equals("BI")).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (rjweeks) [reply src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java]: Removed.

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java [hunk 5]
```diff
@@ -185,21 +188,76 @@ public CaseEvent createCaseEvent(CaseEvent caseEvent, Case newCase, Timestamp ti
...
+    //select * from casesvc.casegroup where partyid = partyId and collectionexerciseid in collectionExercises
+    List<CaseGroup> caseGroups = caseGroupRepo.retrieveByPartyIdInListOfCollEx(partyId, collExs);
+    checkCaseState(category, caseGroups, caseEvent, newCase);
+  }
+
+  /**
+   * This has been triggered by a 'RESPONDENT_ENROLED' event. If we find any associated Case Groups
+   * with only B cases we need to make case 'INACTIONABLE' and create BI case.
+   */
+  private void checkCaseState(final Category category, List<CaseGroup> caseGroups, CaseEvent caseEvent, Case newCase) throws CTPException {
```
[the remark this answers is not in the document]
Comment (rjweeks) [reply src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java]: I've added a comment where it tries to transition case.

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java [hunk 6]
```diff
@@ -185,21 +188,76 @@ public CaseEvent createCaseEvent(CaseEvent caseEvent, Case newCase, Timestamp ti
...
   }
 
+  /**
+   * Use case ID to find case group
+   * Use case group to find collectionEx
+   * Then use CollEx service to find survey ID, then find all collexs for survey
+   * Then get all casge groups for the party ID where Collex ID is in list of collexs
+   * TODO: null checking
+   */
+  private void transitionOtherCaseGroups(final Category category, final CaseEvent caseEvent, final Case targetCase, Case newCase) throws CTPException {
```
[the remark this answers is not in the document]
Comment (rjweeks) [reply src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java]: It's not good practice.  I will raise a ticket to refactor this class.

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java [hunk 7]
```diff
@@ -185,21 +188,76 @@ public CaseEvent createCaseEvent(CaseEvent caseEvent, Case newCase, Timestamp ti
...
       // transition case group status
       transitionCaseGroupStatus(caseEvent, targetCase);
+
+      // if this is a respondent enrolling event
+      if (caseEvent.getCategory().toString().equals("RESPONDENT_ENROLED")) {
+        // are there other case groups that need updating
+        transitionOtherCaseGroups(category, caseEvent, targetCase, newCase);
+      } else {
+        createNewCase(category, caseEvent, targetCase, newCase);
+        effectTargetCaseStateTransition(category, targetCase);
```
[the remark this answers is not in the document]
Comment (rjweeks) [reply src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java]: It's tested ok.

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java [hunk 8]
```diff
@@ -185,21 +188,76 @@ public CaseEvent createCaseEvent(CaseEvent caseEvent, Case newCase, Timestamp ti
...
       // does the event transition the case?
-      effectTargetCaseStateTransition(category, targetCase);
+      //effectTargetCaseStateTransition(category, targetCase);
 
       // should we create an ad hoc action?
       createAdHocAction(category, caseEvent);
 
       // should a new case be created?
-      createNewCase(category, caseEvent, targetCase, newCase);
+      //createNewCase(category, caseEvent, targetCase, newCase);
```
[the remark this answers is not in the document]
Comment (rjweeks) [reply src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java]: Removed.

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseGroupServiceImpl.java [hunk 9]
```diff
@@ -74,4 +88,27 @@ public void transitionCaseGroupStatus(final CaseGroup caseGroup, final CategoryD
...
+   * Use case group to find collectionEx
+   * Then use CollEx service to find survey ID, then find all collexs for survey
+   * Then get all casge groups for the party ID where Collex ID is in list of collexs
+   * TODO: null checking
+   */
+  public List<CaseGroup> transitionOtherCaseGroups(final Category category, final CaseEvent caseEvent, final Case targetCase, Case newCase) throws CTPException {
+    CaseGroup caseGroup = caseGroupRepo.findOne(targetCase.getCaseGroupFK());
+    CollectionExerciseDTO collectionExercise = collectionExerciseSvcClientService
+            .getCollectionExercise(caseGroup.getCollectionExerciseId());
+    // fetch all the collection exercises for a survey
```
Comment (jwgwalton) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseGroupServiceImpl.java]: Comments seem a bit over the top, your variable names seem pretty self documenting.

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java [hunk 10]
```diff
@@ -202,11 +200,43 @@ public CaseEvent createCaseEvent(CaseEvent caseEvent, Case newCase, Timestamp ti
...
+   * This has been triggered by a 'RESPONDENT_ENROLED' event. If we find any associated Case Groups
+   * with only B cases we need to make case 'INACTIONABLE' and create BI case.
+   */
+  private void checkCaseState(final Category category, List<CaseGroup> caseGroups, CaseEvent caseEvent, Case newCase) throws CTPException {
+    // check all case groups are type B. For surveys this is always true
+    List<CaseGroup> caseGroupsToUpdate = caseGroups.stream().filter(cg -> cg.getSampleUnitType().toString().equals("B")).collect(Collectors.toList());
+    for(CaseGroup cg : caseGroupsToUpdate) {
+      // fetch cases associated to case group
+      List<Case> cases = caseRepo.findByCaseGroupFKOrderByCreatedDateTimeDesc(cg.getCaseGroupPK());
+      // find b cases
```
Comment (jwgwalton) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java]: Same on the comments as above.

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java [hunk 11]
```diff
@@ -202,11 +200,43 @@ public CaseEvent createCaseEvent(CaseEvent caseEvent, Case newCase, Timestamp ti
...
+      List<Case> bCases = cases.stream().filter(c -> c.getSampleUnitType().toString().equals("B")).collect(Collectors.toList());
+      // see if any case needs to transition
+      for(Case bCase : bCases) {
+        effectTargetCaseStateTransition(category, bCase);
+        Case c = new Case();
```
Comment (jwgwalton) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java]: I would rename the case to biCase. I was confused until reading the comment for the method.

File: src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java [hunk 12]
```diff
@@ -202,11 +200,43 @@ public CaseEvent createCaseEvent(CaseEvent caseEvent, Case newCase, Timestamp ti
...
     return createdCaseEvent;
   }
 
+  /**
+   * This has been triggered by a 'RESPONDENT_ENROLED' event. If we find any associated Case Groups
+   * with only B cases we need to make case 'INACTIONABLE' and create BI case.
+   */
+  private void checkCaseState(final Category category, List<CaseGroup> caseGroups, CaseEvent caseEvent, Case newCase) throws CTPException {
+    // check all case groups are type B. For surveys this is always true
+    List<CaseGroup> caseGroupsToUpdate = caseGroups.stream().filter(cg -> cg.getSampleUnitType().toString().equals("B")).collect(Collectors.toList());
```
Comment (tej99) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/service/impl/CaseServiceImpl.java]: If this is always true do we need it?

File: src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseGroupRepository.java [hunk 13]
```diff
@@ -18,6 +21,23 @@
...
+    /**
+     * To find CaseGroup by party UUID
+     * @param id the UUID of the Party
+     * @return the matching CaseGroups
+     */
+    List<CaseGroup> findByPartyId(UUID partyId);
+
+    /**
+     * To find CaseGroup by party UUID where Collection Exercise in list supplied
+     * @param id the UUID of the Party
```
Comment (tej99) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseGroupRepository.java]: @param partyId the UUID of the Party

File: src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseGroupRepository.java [hunk 14]
```diff
@@ -18,6 +21,23 @@
...
+     * @param the list of Collection exercises for a survey
```
Comment (tej99) [inline src/main/java/uk/gov/ons/ctp/response/casesvc/domain/repository/CaseGroupRepository.java]: @param collExs list of Collection exercise ids

**Comment by andrewmil** [comment]
[US030: [Required post B&B for CE2] SEFT: View All Collection Exercises For a Survey (add a survey) (XL = 50pts = 10 days)](https://trello.com/c/KEQfg5FA/55-us030-required-post-bb-for-ce2-seft-view-all-collection-exercises-for-a-survey-add-a-survey-xl-50pts-10-days)

**Pull request merged**